src/viam/examples/modules/complex/gizmo/impl.cpp

```diff
@@ -15,20 +15,21 @@ using namespace viam::sdk;
 
 std::string find_arg1(ResourceConfig cfg) {
     auto gizmo_name = cfg.name();
-    auto arg1 = cfg.attributes()->find("arg1");
-    if (arg1 == cfg.attributes()->end()) {
+    auto arg1 = cfg.attributes().find("arg1");
+    if (arg1 == cfg.attributes().end()) {
         std::ostringstream buffer;
         buffer << gizmo_name << ": Required parameter `arg1` not found in configuration";
         throw std::invalid_argument(buffer.str());
     }
-    const auto* const arg1_string = arg1->second->get<std::string>();
-    if (!arg1_string || arg1_string->empty()) {
-        std::ostringstream buffer;
-        buffer << gizmo_name << ": Required non-empty string parameter `arg1`"
-               << "` is either not a string or is an empty string";
-        throw std::invalid_argument(buffer.str());
+
+    const ProtoValue& arg1_val = arg1->second;
+    if (arg1_val.is_a<std::string>() && !arg1_val.get_unchecked<std::string>().empty()) {
+        return arg1_val.get_unchecked<std::string>();
```

Comment:
Is aw this patern in `base/impl.cpp` above too, and sort of shrugged it off, but, it now that I see it twice it feels a little chatty. Is `ProtoValue` missing some affordance for this sort of pattern? Would `std::optional` (or a polyfill for it) simplify things? Should `ProtoValue` have some monadic APIs (and_then, transform, or_else), etc?

This, btw, is what motivated my slack ping about polyfills / C++17 since I feel like `optional` probably plays a role in such an API.

Reply:
Yeah it was bugging me as well that this kept coming up, being that it's the one scenario where we end up more verbose than the previous code. 

I personally am a big fan of optionals and monadic operations but in this case we might be better off with a pointer API. For example Boost.JSON has [these](https://www.boost.org/doc/libs/1_85_0/libs/json/doc/html/json/ref/boost__json__value/if_double.html), and we could spell them as `try_get<T>` or similar, since `if_bool` strikes me as a bit idiosyncratic 

In any case definitely in agreement that this would be nice to simplify, will create a ticket.